### PR TITLE
Heatmap is working with LogScale with invalid values (set to white)

### PR DESCRIPTION
**Pull request by loichuder**
image

After our discussion in #184, I finally settled for the following behaviour:
- Pixels with invalid values are set to white color*
- The store is untouched: the `supportedDomain` is computed in the `HeatmapVis` as the single source of truth and passed to the components needing it.
- As a consequence, `customDomain` and its associated slider do **not** reflect the changes in `supportedDomain`

_* This is the behaviour in matplotlib so I expect users to be used to it:_
image

### CODE REVIEW

File: src/h5web/visualizations/heatmap/hooks.ts [hunk 1]
```diff
@@ -26,20 +27,14 @@ export interface TextureDataState {
...
-      state.customDomain,
-      state.scaleType,
-      state.colorMap,
-    ],
+  const [scaleType, colorMap] = useHeatmapConfig(
+    (state) => [state.scaleType, state.colorMap],
     shallow
   );
 
-  const domain = customDomain || dataDomain;
```
Comment (loichuder) [inline src/h5web/visualizations/heatmap/hooks.ts:41]: `domain` comes from the props now

File: src/h5web/visualizations/heatmap/utils.ts [hunk 2]
```diff
@@ -17,3 +19,28 @@ export function generateCSSLinearGradient(
...
+  }
+
+  const supportedDomain = findDomain(values.filter((x) => x > 0));
+  if (supportedDomain && customDomain) {
+    // Clamp custom domain minimum to the first positive value
+    return [supportedDomain[0], customDomain[1]] as Domain;
+  }
+
+  return supportedDomain;
+}
```
Comment (loichuder) [inline src/h5web/visualizations/heatmap/utils.ts:46]: This is the main point of the MR: adding a function to compute once and for all components needing it, the domain to be used when mapping to colours.

File: src/h5web/visualizations/heatmap/ColorBar.tsx [hunk 3]
```diff
@@ -1,29 +1,28 @@
...
-  );
+import { SCALE_FUNCTIONS, ScaleType, Domain } from '../shared/models';
 
+function ColorBar({
+  domain,
+  scaleType,
+}: {
+  domain: Domain | undefined;
+  scaleType: ScaleType;
+}): JSX.Element {
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/ColorBar.tsx:16]: Formatting can be fixed with:

- `props` + destructuring inside the function
- `Props` interface

File: src/h5web/visualizations/heatmap/ColorBar.tsx [hunk 4]
```diff
@@ -1,29 +1,28 @@
...
-import { useHeatmapConfig } from './config';
 import styles from './HeatmapVis.module.css';
 import { generateCSSLinearGradient } from './utils';
-import { SCALE_FUNCTIONS } from '../shared/models';
-
-function ColorBar(): JSX.Element {
-  const [dataDomain, customDomain, scaleType] = useHeatmapConfig(
-    (state) => [state.dataDomain, state.customDomain, state.scaleType],
-    shallow
-  );
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/ColorBar.tsx:16]: I like the idea of using less of `useHeatmapConfig` so we can pass the config as a prop to `HeatmapVis`. That being said, this will trigger more re-renders of `HeatmapVis` itself, so we may have to add `React.memo()` here and there, like on `VisCanvas` to avoid re-rendering too many children when the user changes the color map or whatever.

File: src/h5web/visualizations/heatmap/HeatmapVis.tsx [hunk 5]
```diff
@@ -21,8 +21,20 @@ function HeatmapVis(props: Props): JSX.Element {
...
+    keepAspectRatio,
+    showGrid,
+  ] = useHeatmapConfig(
+    (state) => [
+      state.dataDomain,
+      state.customDomain,
+      state.scaleType,
+      state.keepAspectRatio,
+      state.showGrid,
+    ],
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/HeatmapVis.tsx:37]: Isn't this basically the whole config now? `const { ... } = useHeatmapConfig()` would be a lot cleaner, I reckon.

File: src/h5web/visualizations/heatmap/hooks.ts [hunk 6]
```diff
@@ -26,20 +27,14 @@ export interface TextureDataState {
...
 ): TextureDataState {
-  const [dataDomain, customDomain, scaleType, colorMap] = useHeatmapConfig(
-    (state) => [
-      state.dataDomain,
-      state.customDomain,
-      state.scaleType,
-      state.colorMap,
-    ],
+  const [scaleType, colorMap] = useHeatmapConfig(
+    (state) => [state.scaleType, state.colorMap],
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/hooks.ts:34]: Might as well pass everything through the props now, don't you think?

File: src/h5web/visualizations/heatmap/utils.ts [hunk 7]
```diff
@@ -17,3 +19,28 @@ export function generateCSSLinearGradient(
...
+  if (!domain) {
+    return undefined;
+  }
+
+  // If the scale is not log or if the domain does not cross zero, the domain does not need to be adjusted
+  if (scaleType !== ScaleType.Log || domain[0] * domain[1] > 0) {
+    return domain;
+  }
+
+  const supportedDomain = findDomain(values.filter((x) => x > 0));
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/utils.ts:39]: This line filters positive values, but the condition above supports datasets with only negative values as well. The `filter` needs a ternary inside it that chooses which function to apply based on whether `domain[0] * domain[1]` is greater or lower than 0. Perhaps `d3-array` has something to simplify all this logic? I haven't looked.

**Review by axelboc — CHANGES_REQUESTED** [review]
Nice and clean 👍  -- just one issue and a couple of minor suggestions.

File: src/h5web/visualizations/heatmap/ColorBar.tsx (hunk 4 above)
Comment (axelboc) [reply src/h5web/visualizations/heatmap/ColorBar.tsx:16]: I've opened an issue to remember.

**Comment by axelboc** [comment]
Oh and I agree with your decisions:
- to show white data points (I believe that's what is done in silx view as well, if I recall)
- to pass the domain through props and reduce the use of `useHeatmapConfig`, as I've mentioned in a comment.

File: src/h5web/visualizations/heatmap/ColorBar.tsx (hunk 3 above)
Comment (loichuder) [reply src/h5web/visualizations/heatmap/ColorBar.tsx:16]: Oops, I wrote that for quick prototyping and let this slide :sweat:

File: src/h5web/visualizations/heatmap/ColorBar.tsx (hunk 4 above)
Comment (loichuder) [reply src/h5web/visualizations/heatmap/ColorBar.tsx:16]: All right. I did not notice any performance drop but it is better to keep it in mind

File: src/h5web/visualizations/heatmap/HeatmapVis.tsx (hunk 5 above)
Comment (loichuder) [reply src/h5web/visualizations/heatmap/HeatmapVis.tsx:37]: Well, there are still the setters inside the config, but we might as well cross the bridge

File: src/h5web/visualizations/heatmap/hooks.ts (hunk 6 above)
Comment (loichuder) [reply src/h5web/visualizations/heatmap/hooks.ts:34]: I think therefore I will do so :slightly_smiling_face:

File: src/h5web/visualizations/heatmap/utils.ts (hunk 7 above)
Comment (loichuder) [reply src/h5web/visualizations/heatmap/utils.ts:39]: I see your point, but I think this necessitates some thought. Example: if you have a dataset with positive and negative values, you may argue that you want the supported domain to be computed from negative values. Then, how do you choose between the two possibilities ?

From my point of view, I expect the logScale to be used only for positive values, which is why I used this condition (and the same in `LineVis`). That being said, `d3-scale` also accepts a domain of negative values so I allow the log scale to take advantage of that for datasets containing only negatives values, with the condition above.

TL;DR: it raises interesting questions that should be addressed in another MR, ideally conjointly with `LineVis`.

**Comment by loichuder** [comment]
> Oh and I fully agree with your decisions:
> 
>     * to show white data points (I believe that's what is done in silx view as well, if I recall)

`silx` uses `matplotlib` so the behaviour should be indeed the same.
> 
>     * to pass the domain through props and reduce the use of `useHeatmapConfig`, as I've mentioned in a comment.
With the removal of the `useInterpolator` hook, the `useHeatmapConfig` is only used in `HeatmapVis` now :+1:

File: src/h5web/visualizations/heatmap/utils.ts (hunk 7 above)
Comment (axelboc) [reply src/h5web/visualizations/heatmap/utils.ts:39]: When you get to line 39, you're certain that the domain does not span both negative and positive values, it's always either one or the other. If you don't use `x < 0` when the domain is entirely negative, then you'll just see a white heatmap.

Comment (axelboc) [reply src/h5web/visualizations/heatmap/utils.ts:39]: > based on whether domain[0] * domain[1] is greater or lower than 0

This was wrong, sorry. You can just check the sign of the min.

File: src/h5web/visualizations/heatmap/HeatmapVis.tsx (hunk 5 above)
Comment (axelboc) [reply src/h5web/visualizations/heatmap/HeatmapVis.tsx]: They can't change so they can't trigger re-renders.

File: src/h5web/visualizations/heatmap/utils.ts (hunk 7 above)
Comment (loichuder) [reply src/h5web/visualizations/heatmap/utils.ts:39]: > When you get to line 39, you're certain that the domain does not span both negative and positive values, it's always either one or the other. 
Hm I don't think I get you there. On the contrary, the early return if `domain[0] * domain[1] > 0` (that is min and max have the same sign) ensures that line 39 is executed *if and only if* the domain spans negative and positive values, which is why the domain needs adjustment.

That being said, the discussion is fruitful as thanks to your comments, I discovered an edge case where the domain is `[-X, 0]` which will not trigger the early return and lead indeed to a white heatmap.

Comment (axelboc) [reply src/h5web/visualizations/heatmap/utils.ts:39]: Oooh right, sorry :sweat: thanks for the clarification - nice catch on the edge case :fishing_pole_and_fish:

**Review by axelboc — APPROVED** [review]
